src/System.Management.Automation/engine/MshMemberInfo.cs

```diff
@@ -2655,136 +2655,87 @@ internal static PSMethod Create(string name, DotNetAdapter dotNetInstanceAdapter
             return method.PSMethodCtor.Invoke(name, dotNetInstanceAdapter, baseObject, method, isSpecial, isHidden);
         }
 
-        static Type GetMethodGroupType(MethodInfo methodInfo)
+        private static Type GetMethodGroupType(MethodInfo methodInfo)
         {
             if (methodInfo.DeclaringType.IsGenericTypeDefinition)
             {
+                // If the method is from a generic type definition, consider it not convertible.
                 return typeof(Func<PSNonBindableType>);
             }
 
             if (methodInfo.IsGenericMethodDefinition)
             {
-                methodInfo = ReplaceGenericTypeArgumentsWithMarkerTypes(methodInfo);
-                if (methodInfo == null)
-                {
-                    // this happens when there are constraints on the generic type parameters
-                    return typeof(Func<PSNonBindableType>);
-                }
+                // For a generic method, it's possible to infer the generic parameters based on the target delegate.
+                // However, we don't yet handle generic methods in PSMethod-to-Delegate conversion, so for now, we
+                // don't produce the metadata type that represents the signature of a generic method.
+                //
+                // Say one day we want to support generic method in PSMethod-to-Delegate conversion and need to produce
+                // the metadata type, we should use the generic parameter types from the MethodInfo directly to construct
+                // the Func<> metadata type. See the concept shown in the following scripts:
+                //    $class = "public class Zoo { public static T GetName<T>(int index, T input) { return default(T); } }"
+                //    Add-Type -TypeDefinition $class
+                //    $method = [Zoo].GetMethod("GetName")
+                //    $allTypes = $method.GetParameters().ParameterType + $method.ReturnType
+                //    $metadataType = [Func`3].MakeGenericType($allTypes)
+                // In this way, '$metadataType.ContainsGenericParameters' returns 'True', indicating it represents a generic method.
+                // And also, given a generic argument type from `$metadataType.GetGenericArguments()`, it's easy to tell if it's a
+                // generic parameter (for example, 'T') based on the property 'IsGenericParameter'.
+                // Moreover, it's also easy to get constraints of the generic parameter, via 'GetGenericParameterConstraints()'
```

Comment:
Much cleaner!

Reply:
We may run into problems with representing methods containing `Span<*>` parameters.
Span cannot be a generic parameter. That is why I had all the convoluted PSGenericType stuff IIRC.

And since span is getting more and more usage, it would be nice if we have a story around it.
Sample below:

```csharp
class A{ public static void Foo(Span<char> x){}}
delegate void SpanAFunc(Span<char> a);
```

```powershell
[SpanAFunc] $x = [A]::Foo;

````

Reply:
But it is probably the right thing to special case `Span` anyway, since it is an exceptional type.

Reply:
Very good point. I will add a placeholder type `PSSpan<>` to represent a `Span<>` type.

Reply:
In fact, it's not just `Span<T>` or `ReadOnlySpan<T>`, but a general problem for all ref-like types, such as `public ref struct Foo { private Span<int> _pointer }`. I don't have a generic solution for the ref-lik type at this moment, but at least we should handle `Span<T>` and `ReadOnlySpan<T>`. I have updated #5618 to track this work.

Interestingly, as `typeof(Func<,>).MakeGenericType(new Type[] { typeof(Span<int>), typeof(int) })` in C# (7.2 language) throws `TypeLoadException` as expected, `[Func`2].MakeGenericType(@([System.Span[int]], [int]))` runs successfully in powershell (both preview.2 and the build from latest master). I'm not sure what is going on with PowerShell.

Good news is that we [catches `TypeLoadException`](https://github.com/PowerShell/PowerShell/pull/6851/files#diff-9bb74f593b0d248c800bb3fe80529244R2709) when constructing the metadata type `Func<>`, so at least a method with ref-like type parameter or return type won't crash PowerShell even if `MakeGenericType` fails -- it will only fails the conversion.

Reply:
I will merge this PR as is, and submit another PR later to address the ref-like types (or at least the `Span<T>` and `ReadOnlySpan<T>`).